pallets/ah-migrator/src/preimage.rs

```diff
@@ -0,0 +1,169 @@
+				}
+
+				match preimage.try_mutate(|p| {
+					p.extend(chunk.chunk_bytes.clone());
+				}) {
+					Some(preimage) => {
+						alias::PreimageFor::<T>::insert(&key, &preimage);
+						preimage
+					},
+					None => {
+						defensive!("Preimage too big");
+						return Err(Error::<T>::TODO);
+					},
+				}
+			},
+			None => {
+				if chunk.chunk_byte_offset != 0 {
+					defensive!("Preimage chunk missing");
+					return Err(Error::<T>::TODO);
+				}
+
+				let preimage: BoundedVec<u8, ConstU32<{ CHUNK_SIZE }>> = chunk.chunk_bytes;
+				debug_assert!(CHUNK_SIZE <= pallet_rc_migrator::preimage::alias::MAX_SIZE);
+				let bounded_preimage: BoundedVec<
+					u8,
+					ConstU32<{ pallet_rc_migrator::preimage::alias::MAX_SIZE }>,
+				> = preimage.into_inner().try_into().expect("Asserted");
+				alias::PreimageFor::<T>::insert(key, &bounded_preimage);
+				bounded_preimage
+			},
+		};
+
+		if preimage.len() == chunk.preimage_len as usize + chunk.chunk_byte_offset as usize {
+			log::debug!(target: LOG_TARGET, "Preimage complete: {}", chunk.preimage_hash);
+		}
+
+		Ok(())
+	}
+
+	pub fn do_receive_preimage_request_statuses(
+		request_status: Vec<RcPreimageRequestStatusOf<T>>,
+	) -> Result<(), Error<T>> {
+		Self::deposit_event(Event::PreimageRequestStatusBatchReceived {
+			count: request_status.len() as u32,
+		});
+		log::info!(target: LOG_TARGET, "Integrating {} preimage request status", request_status.len());
+		let (mut count_good, mut count_bad) = (0, 0);
+
+		for request_status in request_status {
+			match Self::do_receive_preimage_request_status(request_status) {
+				Ok(()) => count_good += 1,
+				Err(e) => {
+					count_bad += 1;
+					log::error!(target: LOG_TARGET, "Error while integrating preimage request status: {:?}", e);
+				},
+			}
+		}
+
+		Self::deposit_event(Event::PreimageRequestStatusBatchProcessed { count_good, count_bad });
+		Ok(())
+	}
+
+	pub fn do_receive_preimage_request_status(
+		mut request_status: RcPreimageRequestStatusOf<T>,
+	) -> Result<(), Error<T>> {
+		if alias::RequestStatusFor::<T>::contains_key(&request_status.hash) {
+			log::warn!(target: LOG_TARGET, "Request status already migrated: {:?}", request_status.hash);
+			return Ok(());
+		}
+
+		let new_ticket = match request_status.request_status {
+			alias::RequestStatus::Unrequested { ticket: (ref who, ref ticket), len } => {
+				let fp = Footprint::from_parts(1, len as usize);
+				ticket.clone().update(&who, fp).ok()
+			},
+			alias::RequestStatus::Requested {
+				maybe_ticket: Some((ref who, ref ticket)),
+				maybe_len: Some(len),
+				..
+			} => {
+				let fp = Footprint::from_parts(1, len as usize);
+				ticket.clone().update(&who, fp).ok()
+			},
+			alias::RequestStatus::Requested { maybe_ticket: Some(_), maybe_len: None, .. } => {
+				defensive!("Ticket cannot be re-evaluated");
+				// I think this is unreachable, but not exactly sure. Either way, nothing that we
+				// could do about it.
+				None
+			},
+			_ => None,
```

Comment:
here we handle `Requested { ticket: None, and maybe_len: Some OR None }`, is it valid?

Reply:
Not exactly sure if it can happen, but we dont need to do anything about it as there is no ticket to update.